Complete the Turbopack migration and remove the paper.js/jsdom resolveAlias workaround

> **Corrected twice after investigation.** The original description blamed Turbopack's handling of paper's `browser` field and listed an App Router migration as a candidate fix — both wrong ([evidence](https://github.com/InformaticsMatters/squonk2-data-manager-ui/issues/2020#issuecomment-5444765488)). The root cause was then narrowed further, and a resolution found ([evidence](https://github.com/InformaticsMatters/squonk2-data-manager-ui/issues/2020#issuecomment-5445264456)). The body below is the current version.

## Context

#2015 moves the build from webpack to Turbopack, because the React Compiler's native Rust port (`experimental.turbopackRustReactCompiler`, Next 16.3) [is only supported under Turbopack and errors outright under webpack](https://nextjs.org/docs/app/api-reference/config/next-config-js/turbopackRustReactCompiler). That switch is only possible today because of a workaround in `next.config.mjs`, and **that workaround is not acceptable as a permanent state** — #2015 should not land while it is load-bearing.

This issue tracks removing it and finishing the migration properly.

## The workaround

```js
turbopack: {
  resolveAlias: {
    "jsdom/lib/jsdom/living/generated/utils": "./src/stubs/jsdom-utils.stub.js",
  },
},
```

It points a real module specifier at a hand-written no-op stub, `src/stubs/jsdom-utils.stub.js`. Neither the alias nor the stub is new — both predate the better-auth migration (`84a93092`), which deleted them at the same time as it pinned `--webpack`. #2015 restores them.

## Root cause

**`paper` requires a package it does not depend on.** Its published manifest declares no `dependencies`, `optionalDependencies` or `peerDependencies` at all — `jsdom` (`^16.7.0`) and `canvas` (`^2.11.2`) are devDependencies only, and devDependencies are never installed for consumers. `dist/node/canvas.js` requires both anyway. The `canvas` require is caught by a `try`/`catch`; the jsdom one is not, so it is resolved statically by any bundler.

`ketcher-core` depends on `paper`, which is how it reaches us. The code never runs — it is guarded at runtime on `canvas`, which is not installed — but the specifier is still resolved at build time, and **nothing in our tree provides that path.**

pnpm is behaving correctly. Under npm/yarn flat hoisting, anything anywhere in the tree pulling jsdom would silently satisfy paper's unqualified require; under pnpm's isolated `node_modules` an undeclared require cannot resolve, so the latent upstream bug is unmaskable.

It only became a build error on the move to Turbopack. Under webpack the sketcher never reached the server compilation; under Turbopack it stays in the SSR graph, where the unresolvable require is fatal. Every failing import trace is an SSR one — the browser graph never fails.

## What has been ruled out

All tested directly, not inferred:

- **App Router does not help.** Reproducing `epam/ketcher`'s `example-ssr` structure here — a `'use client'` module importing ketcher, consumed by a Server Component — fails identically as `[Client Component SSR]`. `'use client'` marks a boundary; it does not remove the module from the server graph.
- **Updating ketcher does not help, and no future release will.** `ketcher-*@3.17.2` hits the same error, and additionally produces 38 unrelated Turbopack errors (`BILN_ALIAS_FORMAT_ERROR_MESSAGE`, `CoreAtom`, `HELM_ALIAS_FORMAT_ERROR_MESSAGE`, …), so a ketcher upgrade is separate work with its own cost. See below for why this is structural rather than a matter of waiting.
- **`serverExternalPackages: ["paper"]` does nothing here.** The alias alone is what is load-bearing.

### Why `epam/ketcher`'s example builds with a clean config

Not a technique, and not something a ketcher release can hand us. `ketcher-core` and `ketcher-react` devDepend on `jest-environment-jsdom` → `jsdom`, which their npm-workspaces layout hoists to the root `node_modules`, where paper's require resolves. **Consumers never install a published package's devDependencies**, so this mechanism is structurally non-transitive — it exists only inside their own checkout. It should not be retested on each ketcher release.

The jsdom version window also matters, and is narrower than first recorded. Checked against the published tarballs — `lib/jsdom/living/generated/utils.js` ships in 16.7.0, 26.1.0 and 27.4.0, and **not** in 28.1.0, 29.1.1 or 30.0.1. It is a jsdom *internal*, not a package export, so its removal in 28 carried no semver obligation and nothing protects the remaining versions.

## Resolution

**A `pnpm` patch on `paper`** — `patches/paper@0.12.18.patch`, one changed line in `dist/node/canvas.js` that assembles the specifier at runtime so bundlers do not resolve it statically. Runtime behaviour is unchanged: the line is only reached when `require('canvas')` succeeds, and installing `canvas` is what brings jsdom into the tree in the first place. Full diff, verification and failure-mode testing in [this comment](https://github.com/InformaticsMatters/squonk2-data-manager-ui/issues/2020#issuecomment-5445264456).

Recorded as considered and rejected:

- **Declare `jsdom` as a devDependency.** Works, and removes both alias and stub — but installs a large tree and pins it to `^27` indefinitely, to satisfy a require that never executes, on an internal path that has already proven it can disappear. Worse than the alias.
- **`pnpm.packageExtensions` on `paper`.** Repairs the declaration rather than the code, which is arguably the more honest fix — but lands in exactly the same place as the above, since it still installs jsdom and still pins the ≤27 range.
- **Keep ketcher and paper off the server graph.** Not pursued. Whether Turbopack should keep an `ssr: false` dynamic import in the SSR graph at all is still an open question, but it is no longer load-bearing for this issue.

Worth doing upstream regardless: paper.js should declare `jsdom` as an `optionalDependency` next to `canvas`, or not require it statically.

## Done when

- `next.config.mjs` carries no hand-written alias for `jsdom/lib/jsdom/living/generated/utils`, and `src/stubs/jsdom-utils.stub.js` is deleted.
- `pnpm build` and `pnpm dev` succeed under Turbopack without it.
- `pnpm lint`, `pnpm tsc` and `pnpm test` pass.

## Resolved by a `pnpm` patch on `paper` — and the root cause is narrower than recorded above

The description frames this as "nothing in our dependency tree provides that path". True, but it stops one step short of the actual defect. The defect is in `paper`, and it is an ordinary undeclared-dependency bug that has nothing to do with Turbopack, Next, or ketcher.

### `paper` declares no runtime dependencies at all

From the installed `paper@0.12.18` manifest:

```
dependencies:         undefined
optionalDependencies: undefined
peerDependencies:     undefined
devDependencies:      jsdom ^16.7.0 | canvas ^2.11.2
```

`dist/node/canvas.js` requires two packages that `paper` never declares for consumers. For `canvas` that is harmless — the require sits in a `try`/`catch`. For `jsdom` it is not caught, so bundlers resolve it statically and it is unresolvable in any tree that does not happen to contain jsdom for unrelated reasons.

pnpm is behaving correctly here. Under npm/yarn flat hoisting, anything anywhere in the tree pulling jsdom would silently satisfy `paper`'s unqualified require. Under pnpm's isolated `node_modules`, an undeclared require cannot resolve. jsdom is nowhere in our tree.

### Why no ketcher release can ever fix this

Above I recorded that `ketcher-*@3.17.2` still fails, tested. Here is the reason, which is stronger than the observation.

Consumers never install a published package's devDependencies. The `jest-environment-jsdom` → `jsdom` edge in `ketcher-core`/`ketcher-react` explains only why **epam's own monorepo checkout** builds: they work in a tree where devDeps *are* installed and npm's flat hoisting parks jsdom at the root, where `paper`'s require finds it. Installing `ketcher-core` from the registry gets you its `dependencies` and nothing else.

So the mechanism that makes their build work is structurally non-transitive. This should not be retested on each ketcher release.

### Correction: the jsdom version window is narrower than stated

The note above says jsdom 26 ships the path and 30 does not. Both true, but the boundary is 27→28. Checked directly against the published tarballs:

| jsdom | ships `lib/jsdom/living/generated/utils.js` |
|---|---|
| 16.7.0 | ✅ |
| 26.1.0 | ✅ |
| 27.4.0 | ✅ |
| 28.1.0 | ❌ |
| 29.1.1 | ❌ |
| 30.0.1 | ❌ |

This matters for avenue 3. The target is a jsdom *internal* — an undocumented deep path, not a package export — so its removal in 28 carried no semver obligation to anyone requiring it this way, and nothing protects the remaining versions either. Note also that `paper` devDepends on jsdom `^16`, so nobody upstream has exercised this code against a modern jsdom at all.

## The fix

`patches/paper@0.12.18.patch`, one changed line in `dist/node/canvas.js`:

```diff
-        idlUtils = require('jsdom/lib/jsdom/living/generated/utils');
+        // The specifier is assembled at runtime so that bundlers do not resolve it statically.
+        // This line is only reached when the optional `canvas` package is installed, which is
+        // what brings jsdom into the tree in the first place; when it is not, the function has
+        // already returned above and this module is never evaluated.
+        idlUtils = require(['jsdom', 'lib', 'jsdom', 'living', 'generated', 'utils'].join('/'));
```

Turbopack cannot constant-fold `.join('/')`, so it stops resolving the specifier. Runtime behaviour is unchanged: the line is only reached when `require('canvas')` succeeds, and installing `canvas` is what puts jsdom in the tree in the first place.

Alongside it: a `patchedDependencies` entry in `pnpm-workspace.yaml` carrying the rationale as a comment, a five-line `pnpm-lock.yaml` diff, `turbopack.resolveAlias` removed from `next.config.mjs`, and `src/stubs/jsdom-utils.stub.js` deleted.

### Verified

- **Reproduced first.** With the alias and stub removed, `pnpm build` fails with the two SSR import traces recorded above.
- **`pnpm build` passes** with the patch and no alias, and emits no Turbopack warning about the now-dynamic require.
- **Runtime confirmed, not assumed.** A temporary `/ketcher-probe` page rendering `SMILESInput` was driven with Playwright against `pnpm dev` (Turbopack): HTTP 200, Ketcher root mounted, benzene rendered from `c1ccccc1`. Probe removed afterwards.
- `pnpm lint` clean, `pnpm tsc` clean, `pnpm test` 808 passed.
- `pnpm install --frozen-lockfile` succeeds, and the file in the `patch_hash=…` store directory carries the patched line.

### The failure mode is safe

Tested by pointing the entry at a non-existent `paper@0.12.99`: pnpm hard-errors with `ERR_PNPM_UNUSED_PATCH`. So if a ketcher bump moves paper off `0.12.18`, install fails loudly rather than silently reverting to a broken build.

## Alternatives, and why not

- **Declare `jsdom` as a devDependency** (avenue 3) — installs a large tree and pins it to `^27` indefinitely, to satisfy a require that never executes, on an internal path that has already proven it can disappear. Worse than the patch, and worse than the alias.
- **`pnpm.packageExtensions` on `paper`** — not previously recorded. It repairs the *declaration* rather than the code, which is arguably the more honest fix for a missing-dependency bug. Not pursued: it lands in exactly the same place as avenue 3, since it still installs jsdom and still pins the ≤27 range. Not tested, on that basis.
- **Keeping ketcher off the server graph** (avenue 1) — not pursued. Still an open question about whether Turbopack should keep an `ssr: false` dynamic import in the SSR graph, but it is no longer load-bearing for this issue.

## Upstreamable

Stated properly, this is a plain bug report for paper.js: `dist/node/canvas.js` requires `jsdom`, which `paper` does not depend on. The fix upstream is to declare it as an `optionalDependency` next to `canvas`, or not to require it statically.

## Status

Built and verified locally on `fix/2020-paper-patch`, branched off #2015's branch. Not yet committed or pushed.

Separately, and out of scope here: `patches/ketcher-standalone.patch` is in the tree from 795e23c3 but nothing references it — no `patchedDependencies` entry, no `patch-package`. Whatever it fixed is currently not being applied.

--- a/AGENTS.md
+++ b/AGENTS.md
@@ -18,7 +18,7 @@ The app has its own git repository (InformaticsMatters/squonk2-data-manager-ui),
 
 ## Commands
 
-- `pnpm dev` — dev server (uses `--webpack`, not Turbopack).
+- `pnpm dev` — dev server (Turbopack).
 - `pnpm build` — production build. Run the separate strict TypeScript and ESLint commands too.
 - `pnpm generate:client:data-manager` / `pnpm generate:client:account-server` — regenerate one local OpenAPI client from its ignored input in `openapi/`.
 - `pnpm generate:clients` — transactionally regenerate both local OpenAPI clients.

--- a/next.config.mjs
+++ b/next.config.mjs
@@ -48,6 +48,13 @@ let nextConfig = {
   // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
   basePath: process.env.NEXT_PUBLIC_BASE_PATH || undefined,
   transpilePackages,
+  // The Pages Router leaves server-side node_modules unbundled, so Turbopack splits
+  // @emotion/react in two: the copy Node loads at runtime for an externalised package, and the
+  // copy bundled into the SSR graph. MUI components then read a different emotion context than
+  // AppCacheProvider writes, fall back to emotion's default "css" cache instead of the "mui"
+  // one, and every server-rendered class name mismatches on hydration. Bundling removes the
+  // boundary, which is what webpack did all along and what the App Router does by default.
+  bundlePagesRouterDependencies: true,
 };
 
 nextConfig = withMDX(nextConfig);

--- a/package.json
+++ b/package.json
@@ -15,9 +15,9 @@
   },
   "scripts": {
     "prepare": "husky",
-    "dev": "next dev --webpack",
-    "dev:debug": "NODE_OPTIONS='--inspect' next dev --webpack",
-    "build": "next build --webpack",
+    "dev": "next dev",
+    "dev:debug": "NODE_OPTIONS='--inspect' next dev",
+    "build": "next build",
     "generate:client:account-server": "node --experimental-strip-types tools/openapi/generate-client.ts account-server",
     "generate:client:data-manager": "node --experimental-strip-types tools/openapi/generate-client.ts data-manager",
     "generate:clients": "node --experimental-strip-types tools/openapi/generate-client.ts all",

